### DISCUSSION TITLE
过期：osc_lazy.lua 功能建议

### DISCUSSION THREAD

**Discussion by dyphire** [hooke007/mpv_PlayKit#35] [回收]
修改版 osc 脚本很好用，不过我遇到一个 title 显示问题需要改进。
之前曾提过某些 rip 组封装的 media-title 信息不是文件信息而是组信息或压制者信息（典型例子：cxcy作品），所以通常情况下使用 filename 显示方案有较高的兼容性。不过存在特殊情况即网页视频，当传递给 mpv 时它会把网址解析为 filename，网页标题为 media-title。

虽然这可以通过指定副标题解决，但临时方案不太优雅；所以是否可以对于 http(magnet/rtmp) 类型地址特殊处理强制显示 media-title？

我目前发现 [playlistmanager.lua](https://github.com/jonniek/mpv-playlistmanager) 和 [SimpleBookmark.lua](https://github.com/Eisa01/mpv-scripts/blob/master/scripts/SimpleBookmark.lua) 脚本实现了此类优化处理，对此参考并修改了个人 [bookmarker_menu.lua](https://github.com/dyphire/MPV-own/blob/master/scripts/bookmarker_menu.lua) 脚本。

但目前 osc 类脚本以及 mpv.net 的 recent 显示功能都未对网址标题解析进行优化，体验有些糟糕。希望能够有办法实现此类优化。

**Comment by hooke007** [comment]
pr welcome

**Reply by dyphire** [reply]
代码白痴流下了无力的泪水。难度太高就算了，本身也不常用 mpv 打开网址，不然也不会现在才发现这个痛点了

顺便问下，是否可以使用快捷键绑定切换 filename 和 media-title 的显示？在使用频率低的情况下似乎这种方案也够用？

**Reply by dyphire** [reply]
暂时选择使用快捷键切换 OSC 显示内容来做临时解决方案了，频率低的情况下足够用了

```
KEY     cycle-values script-opts "osc_lazy-title=${media-title} ${!chapters==0:章节:${chapter}/${chapters}} ${!playlist-count==1:列表:${playlist-pos-1}/${playlist-count}}" "osc_lazy-title=${filename} ${!chapters==0:章节:${chapter}/${chapters}} ${!playlist-count==1:列表:${playlist-pos-1}/${playlist-count}}"
```

**Comment by dyphire** [comment]
又优化了下 [recent.lua](https://github.com/dyphire/MPV-own/blob/master/scripts/recent.lua) 脚本的网址标题解析功能。~~感觉 mpv.net 的 recent 功能实现应该很类似？不知道对汉化版有无帮助。~~ 好吧，我完全看不懂 mpv.net 怎么实现的，请无视我

**Reply by hooke007** [reply]
那个项目几乎没有lua涉及

**Comment by dyphire** [comment]
发现 mpv.net 的 recent 功能和 [recent.lua](https://github.com/dyphire/MPV-own/blob/master/scripts/recent.lua)  脚本的区别了。

recent.lua 脚本在`history.log`文件中记录了 title 和 path 两种数据，可以选择其中一种在 OSD 上显示出来（path 只显示 filename）；而 mpv.net 在`settings.xml`中只记录了 path 数据，因此也只能显示 path。

也就是说 mpv.net 目前的 recent 功能实现方式是无法对 http 类型进行优化处理的，只能说定位上就是本地播放器？

**Comment by dyphire** [comment]
想到一种条件配置写法来实现https协议文件下OSC自动切换 media-title 显示了

```
[protocol.https]
osd-playing-msg = '${media-title} ${!chapters==0:章节:${chapter}/${chapters}} ${!playlist-count==1:列表:${playlist-pos-1}/${playlist-count}}'
script-opts = osc_lazy-title=${media-title} ${!chapters==0:章节:${chapter}/${chapters}} ${!playlist-count==1:列表:${playlist-pos-1}/${playlist-count}}

```

不过怎么改写成用`profile-cond`判定就不懂了

**Reply by dyphire** [reply]
不行，应该需要匹配`protocol-list`才行

**Reply by hooke007** [reply]
写错了
```
[http]
 profile-cond       = string.match(p.path, "(http)")~=nil
 profile-restore    = copy
 osd-playing-msg    = '${media-title} ${!chapters==0:章节:${chapter}/${chapters}} ${!playlist-count==1:列表:${playlist-pos-1}/${playlist-count}}'
 script-opts-append = osc_lazy-title=${media-title} ${!chapters==0:章节:${chapter}/${chapters}} ${!playlist-count==1:列表:${playlist-pos-1}/${playlist-count}}
```

**Reply by dyphire** [reply]
是的，现在可以正常触发了

调整了下触发条件
```
[http]
 profile-desc       = 网络协议下切换显示内容
 profile-cond       = path:match('https?:')~=nil or path:match('magnet:')~=nil or path:match('rtmp:')~=nil
 profile-restore    = copy
 osd-playing-msg    = '${media-title} ${!chapters==0:章节:${chapter}/${chapters}} ${!playlist-count==1:列表:${playlist-pos-1}/${playlist-count}}'
 script-opts-append = osc_lazy-title=${media-title} ${!chapters==0:章节:${chapter}/${chapters}} ${!playlist-count==1:列表:${playlist-pos-1}/${playlist-count}}
```
再次感谢
